### DISCUSSION TITLE
Grace period or early notifications before a streak breaks

### DISCUSSION THREAD

**Discussion by dimo414** [iSoron/uhabits#1923] [Feature Requests]
I have several long-duration habits (e.g. "Clean the bathroom") that I want to do roughly every _n_ days but don't need to be done on an exact cadence. Currently this "works" by breaking the streak and getting a daily notification until I get around to doing the task. It'd be really nice if I could configure a grace period, like "Start notifying me if the streak will break within the next 7 days", or alternatively "A streak is not broken if the task is updated within 7 days of expiration". Sort of similar to the grace period of certain alerting tools like [healthchecks](https://healthchecks.io/docs/configuring_checks/).

There's some precedent for this since daily tasks effectively already have a grace period (you get the notification before expiration), so arguably this would just be making multi-day tasks work more like daily tasks.

**Comment by oxc** [comment]
So basically you're suggesting that the streak will not be (fully) broken if it can still be recovered?

For example, I want to go to the gym 3 times per week, so the first 4 days could be (semi-)filled even if I don't go. If I then do 3 days in a row, the target is reached, but it I miss a single day, the streak will be broken.

**Reply by dimo414** [reply]
Yeah that would be another way to frame it for sure (I actually assumed the 3-times-per-week example worked that way already!).

The crux is that notifications don't start firing until the day the habit is going to break, which makes them not very useful since that day might not happen to be the best day to get the task done. I don't care as much about whether the days are marked semi-filled or not, but I want a longer window of time than one day to act on a habit once the streak is about to break.

As another example, I might want a "call my mom every two weeks" habit to start notifying me when I'm within four days of breaking the streak so I know with some lead time that I need to do that, even if I can't do it today.

**Comment by I-Dont-Remember** [comment]
Love this feature idea! Would be great for Habits to have a smooth way to handle the "flexible" recurring chores of life, like the example of cleaning. There are lots of "life maintenance" actions we need to take a on a roughly consistent cadence, but in my experience most todo list apps/habit trackers don't handle them well.

It would be cool to have flexible "due date" habits have some concept of HOW expired it is. _The following format is just to get ideas flowing, not saying this is how it should work in the app._

E.g. if I haven't cleaned the bathroom in 4 days after my 2 week interval, it's "yellow/warning" 🟡 . But if it's been 2 "cycles" (aka 4 weeks past), then it's "red/bad" 🔴 and should be prioritized.
